src/PublicApiGenerator/CodeTypeReferenceBuilder.cs

```diff
@@ -0,0 +1,142 @@
+                return CreateCodeTypeReferenceWithNullabilityMap(genericArgs.Single(), nullabilityMap, true);
+            }
+            else
+            {
+                return new CodeTypeReference(typeName, CreateGenericArguments(type, nullabilityMap));
+            }
+        }
+
+        private static IEnumerator<bool> GetNullabilityMap(ICustomAttributeProvider attributeProvider)
+        {
+            if (attributeProvider is SuppressNullableProvider)
+                return Enumerable.Repeat(false, MAX_COUNT).GetEnumerator();
+
+            var nullableAttr = attributeProvider?.CustomAttributes.SingleOrDefault(d => d.AttributeType.FullName == "System.Runtime.CompilerServices.NullableAttribute");
+
+            if (nullableAttr == null)
+            {
+                foreach (var provider in NullableContext.Providers)
+                {
+                    nullableAttr = provider.CustomAttributes.SingleOrDefault(d => d.AttributeType.FullName == "System.Runtime.CompilerServices.NullableContextAttribute");
+                    if (nullableAttr != null)
+                        break;
+                }
+            }
+
+            if (nullableAttr == null)
+                return Enumerable.Repeat(false, MAX_COUNT).GetEnumerator();
+
+            var value = nullableAttr.ConstructorArguments[0].Value;
+            if (value is CustomAttributeArgument[] arguments)
+                return arguments.Select(a => (byte)a.Value == 2).GetEnumerator();
+
+            return Enumerable.Repeat((byte)value == 2, MAX_COUNT).GetEnumerator();
+        }
+
+        // The compiler optimizes the size of metadata bypassing a sequence of bytes for value types.
+        // Thus, it can even delete the entire NullableAttribute if the whole signature consists only of value types,
+        // for example KeyValuePair<int, int?>, thus we can call IsNullable() only by looking first deep into the signature
+        private static bool HasAnyReferenceType(TypeReference type)
+        {
+            if (!type.IsValueType)
+                return true;
+
+            // ReSharper disable once RedundantEnumerableCastCall
+            var genericArgs = type is IGenericInstance instance ? instance.GenericArguments : type.HasGenericParameters ? type.GenericParameters.Cast<TypeReference>() : null;
+            if (genericArgs == null) return false;
+
+            foreach (var argument in genericArgs)
+            {
+                if (HasAnyReferenceType(argument))
+                    return true;
+            }
+
+            return false;
+        }
+
+        static string GetTypeName(TypeReference type, IEnumerator<bool> nullabilityMap, bool forceNullable)
+        {
+            bool nullable = forceNullable || HasAnyReferenceType(type) && IsNullable();
+
+            var typeName = GetTypeNameCore(type, nullabilityMap, nullable);
+
+            if (nullable)
+                typeName = CSharpAlias.Get(typeName) + "?";
```

Comment:
Kudos to keep it consistent with the other string concatenations and not use interpolation yet.